packages/eslint-plugin/src/util/isTypeUnchanged.ts

```diff
@@ -0,0 +1,39 @@
+import { isTypeFlagSet } from '@typescript-eslint/type-utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+export function isTypeUnchanged(
+  compilerOptions: ts.CompilerOptions,
+  uncast: ts.Type,
+  cast: ts.Type,
+): boolean {
+  if (uncast === cast) {
+    return true;
+  }
+
+  if (
+    isTypeFlagSet(uncast, ts.TypeFlags.Undefined) &&
+    isTypeFlagSet(cast, ts.TypeFlags.Undefined) &&
+    tsutils.isCompilerOptionEnabled(
+      compilerOptions,
+      'exactOptionalPropertyTypes',
+    )
+  ) {
+    const uncastParts = tsutils
+      .unionTypeParts(uncast)
+      .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+    const castParts = tsutils
+      .unionTypeParts(cast)
+      .filter(part => !isTypeFlagSet(part, ts.TypeFlags.Undefined));
+
+    if (uncastParts.length !== castParts.length) {
+      return false;
+    }
+
+    const uncastPartsSet = new Set(uncastParts);
+    return castParts.every(part => uncastPartsSet.has(part));
+  }
```

Comment:
I'm not sure if this is actually necessary for what I'm using this for, and this edge case currently isn't tested on `no-unsafe-type-assertion.test.ts`. I'll be happy to either update this or cover this with a test if the recent changes make sense.

Reply:
I think if there's a test case that passes only with this logic, then yeah, let's use it. Otherwise just a `===` is sufficient IMO.

Great extract+reuse!

Reply:
Thanks! I've reverted the extract and reuse to stick with a simple `===` check. If I see an edge case in which it's necessary I'll add it back 👍